File: frontend/src/api/carbonNews.js
/** 碳新闻：实时爬取 + 一问一答 */
import { api, IMPORT_API_TIMEOUT_MS } from "./http.js";

/** 一次爬取最多 100 条详情，耗时较长 */
const CRAWL_TIMEOUT_MS = 600_000;

export async function crawlCarbonNews({
  keyword = "碳",
  limit = 100,
} = {}) {
  return api("/api/v1/carbon-news/crawl", {
    method: "POST",
    body: JSON.stringify({ keyword, limit }),
    timeoutMs: CRAWL_TIMEOUT_MS,
  });
}

export async function askCarbonNews({ question, items }) {
  return api("/api/v1/carbon-news/ask", {
    method: "POST",
    body: JSON.stringify({ question, items }),
    timeoutMs: IMPORT_API_TIMEOUT_MS,
  });
}

File: backend/app/api/carbon_news.py
"""碳新闻 API：实时爬取 + 基于当次结果的一问一答。"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException

from app.api.deps import get_current_user, require_feature
from app.models.org import User
from app.schemas.carbon_news import (
    CarbonNewsAskIn,
    CarbonNewsAskOut,
    CarbonNewsCrawlIn,
    CarbonNewsCrawlOut,
)
from app.schemas.common import ApiResponse
from app.services import carbon_news_service as svc

router = APIRouter(
    prefix="/carbon-news",
    tags=["carbon-news"],
    dependencies=[Depends(require_feature("carbon_news"))],
)


@router.post("/crawl", response_model=ApiResponse[CarbonNewsCrawlOut])
async def crawl_carbon_news(
    body: CarbonNewsCrawlIn,
    _: Annotated[User, Depends(get_current_user)],
) -> ApiResponse[CarbonNewsCrawlOut]:
    try:
        data = await svc.crawl_carbon_news(
            keyword=body.keyword,
            limit=body.limit,
        )
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"爬取失败：{exc}") from exc
    return ApiResponse(data=data)


@router.post("/ask", response_model=ApiResponse[CarbonNewsAskOut])
async def ask_carbon_news(
    body: CarbonNewsAskIn,
    _: Annotated[User, Depends(get_current_user)],
) -> ApiResponse[CarbonNewsAskOut]:
    question = (body.question or "").strip()
    if not question:
        raise HTTPException(status_code=400, detail="question 不能为空")
    data = await svc.ask_carbon_news(question=question, items=list(body.items or []))
    return ApiResponse(data=data)
